sample-tests-and-e2e-tests/src/use-effect/use-effect-component-with-props.test.jsx

```diff
@@ -0,0 +1,155 @@
+
+import React from 'react'
+import { shallow } from 'enzyme';
+import { act } from 'react-dom/test-utils';
+
+import { expensiveOperation1, expensiveOperation2, expensiveOperation3, expensiveOperation4 } from './expensiveOperations'
+
+import ComponentWithProps from './use-effect-component-with-props'
+
+jest.mock('./expensiveOperations', () => ({
+    expensiveOperation1: jest.fn(),
+    expensiveOperation2: jest.fn(),
+    expensiveOperation3: jest.fn(),
+    expensiveOperation4: jest.fn(),
+}))
+
+describe('Component with props', () => {
+    const originalString = "A String"
+    const originalArray = ["An", "Array", "Of", "Strings"]
+
+    let wrapper = shallow(<ComponentWithProps strProp={originalString} arrProp={originalArray}/>);
+
+    beforeEach(() => {
+        expensiveOperation1.mockClear();
+        expensiveOperation2.mockClear();
+        expensiveOperation3.mockClear();
+        expensiveOperation4.mockClear();
+
+        wrapper = shallow(<ComponentWithProps strProp={originalString} arrProp={originalArray}/>);
+    })
+
+    it("Should create and call expensive operation 1", () => {
+        expect(expensiveOperation1).toHaveBeenCalledTimes(1);
+    })
+
+    it("Should create and call expensive operation 2", () => {
+        expect(expensiveOperation2).toHaveBeenCalledTimes(1);
+    })
+
+    it("Should create and call expensive operation 3", () => {
+        expect(expensiveOperation3).toHaveBeenCalledTimes(1);
+    })
+
+    it("Should create and call expensive operation 4", () => {
+        expect(expensiveOperation4).toHaveBeenCalledTimes(1);
+    })
+
+    it("Should render a button", () => {
```

Comment:
Do we need this test? It doesn't depend on `useEffect()` at all.

Reply:
> Hi @rcbevans ,
> 
> First of all, thank you so much for going through the code and submitting the pull request.
> 
> I do hope that you find the library useful.
> 
> I really like your additions to the unit test. They'll guarantee that the library does work with reference types.
> 
> What are your thoughts behind the new e2e tests? I like that they make it explicit when `useEffect()` is triggered, a.k.a on property changes. But that can be also viewed as a bit of a repetition over the unit tests? After all, we already tests there when `useEffect()` is called and when it's not.

You're welcome.

My motivation for adding the tests was that a colleague attempted to use the library in a nest test suite the other day and it didn't appear to work; the effect was being re-executed even when the ref dep (shouldn't have) changed.  These tests were largely an exercise to prove to ourselves that the problem with those failing tests was actually a problem elsewhere, and not a consequence of these mocks.

As for the 'unrelated' tests, it's true that they don't rely on use effect, but it's largely because I wrote the tests progressively and wanted to assert that my test runner was seeing things as I expected it to, before moving onto testing the behavior I was unsure about.

I'm happy to go through and remove tests not explicitly verifying expectations for use effect, and consolidate multiple test cases that utilize the same trigger of the re-render if you prefer that testing style.

Reply:
@rcbevans, 

That makes sense! Let's bring your tests it. That's the whole purpose of this sample e2e project. So I'm glad it's helping and that you're contributing.

Just one minor thing, is it possible to use two spaces for indentations and add semi-colons at the end of each statement? I apologise for not using Prettier. I'll add it later.

And then we merge :).